+htp #21340 Add withMaterializer to Java DSL Unmarshaller

jypma opened the pull request:
This gives access to the `Materializer` for custom unmarshallers, and creates a convenience `entityToSource()` unmarshaller.

Also repairs `Marshaller.oneOf` with 2 args, which Java wouldn't resolve properly.

Fixes https://github.com/akka/akka/issues/21340

ktoso commented on akka-http/src/main/scala/akka/http/javadsl/marshalling/Marshaller.scala:
```diff
@@ -88,8 +88,8 @@ object Marshaller {
    */
-  def oneOf[A, B](ms: Marshaller[A, B]*): Marshaller[A, B] = {
-    fromScala(marshalling.Marshaller.oneOf[A, B](ms.map(_.asScala): _*))
+  def oneOf[A, B](m1: Marshaller[A, B], m2: Marshaller[A, B]): Marshaller[A, B] = {
```
why this change? Previous was more general

jypma replied:
It didn't work, compiler/interop issue, see the existing comment above. I think the other n-arity methods were introduced for the same reason, and this line was perhaps simply forgotten?

ktoso replied:
Oh right, it should be: 

```
import scala.annotation.varargs

@varargs
def oneOf[A, B](ms: Marshaller[A, B]*): Marshaller[A, B] = {
```

Please fix this way instead :)

jypma replied:
In Java you'd then get the "heap pollution" warning at call sites, afair. I don't think I could reliably remove that in Java code, let alone scala.

My current version:
```scala
  @SafeVarargs @varargs def oneOf[A, B](ms: Marshaller[A, B]*): Marshaller[A, B] = {
    fromScala(marshalling.Marshaller.oneOf[A, B](ms.map(_.asScala): _*))
  }
```
generates compiler warnings at all call sites:
```
[warn] ..../akka-http-tests/src/test/java/akka/http/javadsl/server/JavaRouteTest.java:78: unchecked generic array creation for varargs parameter of type akka.http.javadsl.marshalling.Marshaller<java.util.UUID,akka.http.javadsl.model.RequestEntity>[]
[warn]     Marshaller.oneOf(
```
A (small) advantage of the n-arity methods would be that we can introduce type variance...

ktoso commented on akka-http/src/main/scala/akka/http/javadsl/unmarshalling/Unmarshaller.scala:
```diff
@@ -87,6 +95,9 @@ object Unmarshaller {
   }
 
+  def withMaterializer[A, B](f: Function3[ExecutionContext, Materializer, A, CompletionStage[B]]): Unmarshaller[A, B] =
+    unmarshalling.Unmarshaller.withMaterializer(ctx ⇒ mat ⇒ a ⇒ f.apply(ctx, mat, a).toScala)
```
good

jypma committed:
+htp #21340 Add withMaterializer to Java DSL Unmarshaller

This gives access to the Materializer for custom unmarshallers, and
creates a convenience entityToSource() unmarshaller.

Also repairs Marshaller.oneOf with 2 args, which Java wouldn't resolve
properly.

jypma commented:
@ktoso can you reproduce the compiler warnings for references to `Marshaller.oneOf`?

ktoso commented:
Focused on a different PR for now, please be patient for a bit.

jypma commented:
n/p

ktoso commented:
please review @jrudolph

ktoso commented:
IMO LGTM, I'm ok with the warnings I guess if we can't supress them (weird though)

ktoso commented:
PLS BUILD

jypma commented:
@ktoso I would NOT be ok with the warnings. It means that as a Java developer, having a class that uses this DSL will always have at least one compiler warning on it, unless I selectively turn off warnings manually.

The origin of "weird" is that arrays in Java were contravariant long before generics were introduced, and the two type systems are incompatible. Because of that, arrays and generics (and, hence, varargs and generics) are generally not mixed in Java.

Maybe a compromise is to keep the, say, up to 4-ary methods, and the vargs as well? Compiler should pick the specific ones over the varargs.

johanandren commented:
This PR is moved to akka-http, can we close it?

ktoso commented:
The PR wasn't moved, an issue from this PR was created. 
I'm not completely sure what we want to do about this one, WDYT @jrudolph ?

ktoso commented:
@jypma would you be able to move it over to http://github.com/akka/akka-http ?
Do you have an idea how to solve the warnings issue?
